Build fails with `NullPointerException` during discovery phase

## Overview

While trying to migrate the Spring Framework build to the TestNG `TestEngine` (see https://github.com/spring-projects/spring-framework/issues/27406), I encounter a `NullPointerException` which fails the build.

## Exception

```
org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 168.
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: org.junit.platform.commons.JUnitException: TestEngine with ID 'testng' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:160)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:134)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:80)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:110)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	... 18 more
Caused by: java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011)
	at java.util.concurrent.ConcurrentHashMap$KeySetView.add(ConcurrentHashMap.java:4595)
	at org.junit.support.testng.engine.TestClassRegistry.lambda$start$0(TestClassRegistry.java:29)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
	at org.junit.support.testng.engine.TestClassRegistry.start(TestClassRegistry.java:27)
	at org.junit.support.testng.engine.DiscoveryListener.onBeforeClass(DiscoveryListener.java:37)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:167)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.testng.TestRunner.privateRun(TestRunner.java:794)
	at org.testng.TestRunner.run(TestRunner.java:596)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:377)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:371)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:332)
	at org.testng.SuiteRunner.run(SuiteRunner.java:276)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1212)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1134)
	at org.testng.TestNG.runSuites(TestNG.java:1063)
	at org.testng.TestNG.run(TestNG.java:1031)
	at org.junit.support.testng.engine.TestNGTestEngine.configureAndRun(TestNGTestEngine.java:182)
	at org.junit.support.testng.engine.TestNGTestEngine.lambda$discover$1(TestNGTestEngine.java:91)
	at org.junit.support.testng.engine.TestNGTestEngine.withTemporarySystemProperty(TestNGTestEngine.java:190)
	at org.junit.support.testng.engine.TestNGTestEngine.discover(TestNGTestEngine.java:90)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:152)
	... 29 more
```

## Example

Executing `gradle :spring-test:check` in the following branch reproduces the issue.

- https://github.com/sbrannen/spring-framework/tree/issues/gh-27406-TestNG-TestEngine

--- a/src/main/java/org/junit/support/testng/engine/DiscoveryListener.java
+++ b/src/main/java/org/junit/support/testng/engine/DiscoveryListener.java
@@ -12,8 +12,12 @@ package org.junit.support.testng.engine;
 
 import java.util.HashSet;
 import java.util.Set;
+import java.util.function.Predicate;
 
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.Filter;
 import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.discovery.ClassNameFilter;
 import org.testng.ITestClass;
 import org.testng.ITestResult;
 
@@ -21,8 +25,10 @@ class DiscoveryListener extends DefaultListener {
 
 	private final TestClassRegistry testClassRegistry = new TestClassRegistry();
 	private final TestNGEngineDescriptor engineDescriptor;
+	private final Predicate<String> classNameFilter;
 
-	public DiscoveryListener(TestNGEngineDescriptor engineDescriptor) {
+	public DiscoveryListener(EngineDiscoveryRequest request, TestNGEngineDescriptor engineDescriptor) {
+		this.classNameFilter = Filter.composeFilters(request.getFiltersByType(ClassNameFilter.class)).toPredicate();
 		this.engineDescriptor = engineDescriptor;
 	}
 
@@ -34,8 +40,15 @@ class DiscoveryListener extends DefaultListener {
 
 	@Override
 	public void onBeforeClass(ITestClass testClass) {
-		testClassRegistry.start(testClass.getRealClass(),
-			() -> engineDescriptor.findClassDescriptor(testClass.getRealClass()));
+		testClassRegistry.start(testClass.getRealClass(), realClass -> {
+			ClassDescriptor classDescriptor = engineDescriptor.findClassDescriptor(realClass);
+			if (classDescriptor == null && classNameFilter.test(realClass.getName())) {
+				classDescriptor = engineDescriptor.getTestDescriptorFactory().createClassDescriptor(engineDescriptor,
+					realClass);
+				engineDescriptor.addChild(classDescriptor);
+			}
+			return classDescriptor;
+		});
 	}
 
 	@Override

--- a/src/main/java/org/junit/support/testng/engine/ExecutionListener.java
+++ b/src/main/java/org/junit/support/testng/engine/ExecutionListener.java
@@ -55,7 +55,7 @@ class ExecutionListener extends DefaultListener {
 	public void onBeforeClass(ITestClass testClass) {
 		ClassDescriptor classDescriptor = requireNonNull(engineDescriptor.findClassDescriptor(testClass.getRealClass()),
 			"Missing class descriptor");
-		testClassRegistry.start(testClass.getRealClass(), () -> {
+		testClassRegistry.start(testClass.getRealClass(), __ -> {
 			delegate.executionStarted(classDescriptor);
 			return classDescriptor;
 		});

--- a/src/main/java/org/junit/support/testng/engine/TestClassRegistry.java
+++ b/src/main/java/org/junit/support/testng/engine/TestClassRegistry.java
@@ -16,20 +16,25 @@ import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Consumer;
-import java.util.function.Supplier;
+import java.util.function.Function;
 
 class TestClassRegistry {
 
 	private final Set<ClassDescriptor> classDescriptors = ConcurrentHashMap.newKeySet();
 	private final Map<Class<?>, Entry> testClasses = new ConcurrentHashMap<>();
 
-	void start(Class<?> testClass, Supplier<ClassDescriptor> onFirst) {
+	void start(Class<?> testClass, Function<Class<?>, ClassDescriptor> onFirst) {
 		Entry entry = testClasses.computeIfAbsent(testClass, __ -> {
-			ClassDescriptor classDescriptor = onFirst.get();
-			classDescriptors.add(classDescriptor);
-			return new Entry(classDescriptor);
+			ClassDescriptor classDescriptor = onFirst.apply(testClass);
+			if (classDescriptor != null) {
+				classDescriptors.add(classDescriptor);
+				return new Entry(classDescriptor);
+			}
+			return null;
 		});
-		entry.inProgress.incrementAndGet();
+		if (entry != null) {
+			entry.inProgress.incrementAndGet();
+		}
 	}
 
 	Optional<ClassDescriptor> get(Class<?> testClass) {

--- a/src/main/java/org/junit/support/testng/engine/TestNGTestEngine.java
+++ b/src/main/java/org/junit/support/testng/engine/TestNGTestEngine.java
@@ -84,7 +84,7 @@ public class TestNGTestEngine implements TestEngine {
 		List<String> methodNames = engineDescriptor.getQualifiedMethodNames();
 
 		ConfigurationParameters configurationParameters = request.getConfigurationParameters();
-		DiscoveryListener listener = new DiscoveryListener(engineDescriptor);
+		DiscoveryListener listener = new DiscoveryListener(request, engineDescriptor);
 
 		if (testClasses.length > 0) {
 			withTemporarySystemProperty(TESTNG_MODE_DRYRUN, "true",

--- a/src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java
+++ b/src/test/java/org/junit/support/testng/engine/DiscoveryIntegrationTests.java
@@ -32,6 +32,7 @@ import example.basics.IgnoredTestCase;
 import example.basics.InheritedClassLevelOnlyAnnotationTestCase;
 import example.basics.InheritingSubClassTestCase;
 import example.basics.JUnitTestCase;
+import example.basics.NestedTestClass;
 import example.basics.SimpleTestCase;
 import example.basics.SuccessPercentageTestCase;
 import example.basics.TwoMethodsTestCase;
@@ -285,6 +286,32 @@ class DiscoveryIntegrationTests extends AbstractIntegrationTests {
 		var rootDescriptor = testEngine.discover(request, engineId);
 
 		assertThat(rootDescriptor.getChildren()).isEmpty();
+	}
+
+	@Test
+	void discoversNestedTestClasses() {
+		var selectedTestClass = NestedTestClass.class;
+		var request = request().selectors(selectClass(selectedTestClass)).build();
+
+		var rootDescriptor = testEngine.discover(request, engineId);
+
+		assertThat(rootDescriptor.getUniqueId()).isEqualTo(engineId);
+		assertThat(rootDescriptor.getChildren()).hasSize(2);
+
+		Map<String, TestDescriptor> classDescriptors = rootDescriptor.getChildren().stream() //
+				.collect(toMap(TestDescriptor::getDisplayName, identity()));
+
+		TestDescriptor classDescriptor = classDescriptors.get("A");
+		assertThat(classDescriptor.getLegacyReportingName()).isEqualTo(NestedTestClass.A.class.getName());
+		assertThat(classDescriptor.getType()).isEqualTo(CONTAINER);
+		assertThat(classDescriptor.getSource()).contains(ClassSource.from(NestedTestClass.A.class));
+		assertThat(classDescriptor.getChildren()).hasSize(1);
+
+		classDescriptor = classDescriptors.get("B");
+		assertThat(classDescriptor.getLegacyReportingName()).isEqualTo(NestedTestClass.B.class.getName());
+		assertThat(classDescriptor.getType()).isEqualTo(CONTAINER);
+		assertThat(classDescriptor.getSource()).contains(ClassSource.from(NestedTestClass.B.class));
+		assertThat(classDescriptor.getChildren()).hasSize(1);
 	}
 
 	interface InterfaceTestCase {

--- a/src/test/java/org/junit/support/testng/engine/ReportingIntegrationTests.java
+++ b/src/test/java/org/junit/support/testng/engine/ReportingIntegrationTests.java
@@ -13,6 +13,7 @@ package org.junit.support.testng.engine;
 import static org.junit.platform.commons.util.StringUtils.isBlank;
 import static org.junit.platform.engine.FilterResult.excluded;
 import static org.junit.platform.engine.FilterResult.includedIf;
+import static org.junit.platform.engine.discovery.ClassNameFilter.excludeClassNamePatterns;
 import static org.junit.platform.engine.discovery.DiscoverySelectors.selectClass;
 import static org.junit.platform.engine.discovery.DiscoverySelectors.selectMethod;
 import static org.junit.platform.engine.discovery.DiscoverySelectors.selectUniqueId;
@@ -35,6 +36,7 @@ import java.util.Map;
 import example.basics.CustomAttributeTestCase;
 import example.basics.ExpectedExceptionsTestCase;
 import example.basics.InheritingSubClassTestCase;
+import example.basics.NestedTestClass;
 import example.basics.ParallelExecutionTestCase;
 import example.basics.RetriedTestCase;
 import example.basics.SimpleTestCase;
@@ -46,6 +48,7 @@ import example.dataproviders.DataProviderMethodTestCase;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
+import org.junit.platform.engine.Filter;
 import org.junit.platform.engine.UniqueId;
 import org.junit.platform.engine.support.descriptor.MethodSource;
 import org.junit.platform.launcher.PostDiscoveryFilter;
@@ -300,6 +303,19 @@ class ReportingIntegrationTests extends AbstractIntegrationTests {
 			event(test("method:test()"), dynamicTestRegistered("invoc")), //
 			event(container("method:test()"), finishedSuccessfully()), //
 			event(testClass(testClass), finishedSuccessfully()));
+	}
+
+	@Test
+	void onlyExecutesNestedTestClassesThatMatchClassNameFilter() {
+		var selectedTestClass = NestedTestClass.class;
+
+		var results = testNGEngine() //
+				.selectors(selectClass(selectedTestClass)) //
+				.filters((Filter<?>) excludeClassNamePatterns(".*A$")) //
+				.execute();
+
+		results.containerEvents().assertStatistics(stats -> stats.started(2).finished(2));
+		results.testEvents().assertStatistics(stats -> stats.started(1).finished(1));
 	}
 
 }

--- a/src/testFixtures/java/example/basics/NestedTestClass.java
+++ b/src/testFixtures/java/example/basics/NestedTestClass.java
@@ -1,0 +1,28 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package example.basics;
+
+import org.testng.annotations.Test;
+
+public class NestedTestClass {
+
+	public static class A {
+		@Test
+		public void test() {
+		}
+	}
+
+	public static class B {
+		@Test
+		public void test() {
+		}
+	}
+}

